src/batch/src/executor2/join/nested_loop_join.rs

```diff
@@ -89,83 +91,67 @@ enum NestedLoopJoinState {
     Done,
 }
 
-#[async_trait::async_trait]
-impl Executor for NestedLoopJoinExecutor {
-    async fn open(&mut self) -> Result<()> {
-        match self.state {
-            NestedLoopJoinState::Build => {
-                // TODO: Do not all read to memory if OOM.
-                self.build_table.load_data().await?;
-                self.state = NestedLoopJoinState::FirstProbe;
-            }
-            _ => unreachable!(),
-        }
-        Ok(())
+impl Executor2 for NestedLoopJoinExecutor2 {
+    fn schema(&self) -> &Schema {
+        &self.schema
     }
 
-    async fn next(&mut self) -> Result<Option<DataChunk>> {
-        if let Some(last_chunk) = self.last_chunk.take() {
-            let (left_data_chunk, return_chunk) = self.chunk_builder.append_chunk(last_chunk)?;
-            self.last_chunk = left_data_chunk;
-            if let Some(return_chunk_inner) = return_chunk {
-                return Ok(Some(return_chunk_inner));
-            }
-        }
-        // Make sure all buffered chunk are flushed before generate more chunks.
-        // It's impossible that: left chunk not null and return chunk is null.
-        assert!(self.last_chunk.is_none());
+    fn identity(&self) -> &str {
+        &self.identity
+    }
 
-        loop {
-            match self.state {
-                NestedLoopJoinState::Build => unreachable!(),
+    fn execute(self: Box<Self>) -> BoxedDataChunkStream {
+        self.do_execute()
+    }
+}
 
+impl NestedLoopJoinExecutor2 {
+    #[try_stream(boxed, ok = DataChunk, error = RwError)]
+    async fn do_execute(mut self: Box<Self>) {
+        let mut state = NestedLoopJoinState::Build;
+        loop {
+            match state {
```

Comment:
@liurenjie1024 I think the[ previous hash join implement](https://github.com/singularity-data/risingwave/pull/2119/files#r861413371)  is not very good, this change retains the match control flow is more intuitive, which one do you think is better